### PR TITLE
[Merged by Bors] - feat(LinearAlgebra/BilinearForm/TensorProduct): base change of bilinear forms

### PR DESCRIPTION
**Pull request by eric-wieser**
This generalizes the existing `BilinForm.tensorDistrib` to be heterogenous in the rings it uses, such that a base change,
```lean
protected def baseChange (B : BilinForm R M₂) :
  BilinForm A (A ⊗[R] M₂) :=
```
falls out as a special case. I do not attempt to generalize `BilinForm.tensorDistribEquiv`.

Unfortunately, this changes the defeq as
```diff
-(B₁.tmul B₂) (m₁ ⊗ₜ m₂) (m₁' ⊗ₜ m₂') = B₁ m₁ m₁' * B₂ m₂ m₂'
+(B₁.tmul B₂) (m₁ ⊗ₜ m₂) (m₁' ⊗ₜ m₂') = B₂ m₂ m₂' • B₁ m₁ m₁'
```
We could fix this by using the right action instead, but that's a lot of work for a very minorly annoying defeq.

This also breaks the defeq of `tensorDistribEquiv B = tensorDistrib B`; though the reason is more complicated than the scalar action issue above. It would be fixed if we defined all the homogenous operations on tensor products as special cases of the heterogenous ones, but that's also a lot of work for a very small win.

This is a port of work from https://github.com/pygae/lean-ga/pull/31, and almost at the end of the path to a base change of quadratic forms and clifford algebras.

This was independently developed at the Leiden workshop as [`BilinForm.baseChange`](https://github.com/alexjbest/ant-lorentz/blob/1f97add294b2d50f99537c15583666d78b0d7e24/AntLorentz/BaseChange.lean#L75-L85), though the results there are not sorry-free.

---

- [x] depends on: #6035
- [x] depends on: #6209
- [x] depends on: #6417

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/LinearAlgebra/BilinearForm/TensorProduct.lean [hunk 1]
```diff
@@ -21,52 +22,75 @@ import Mathlib.LinearAlgebra.TensorProduct
...
-def tensorDistrib : BilinForm R M₁ ⊗[R] BilinForm R M₂ →ₗ[R] BilinForm R (M₁ ⊗[R] M₂) :=
-  ((TensorProduct.tensorTensorTensorComm R _ _ _ _).dualMap ≪≫ₗ
-    (TensorProduct.lift.equiv R _ _ _).symm ≪≫ₗ LinearMap.toBilin).toLinearMap ∘ₗ
-  TensorProduct.dualDistrib R _ _ ∘ₗ
-  (TensorProduct.congr (BilinForm.toLin ≪≫ₗ TensorProduct.lift.equiv R _ _ _)
+variable [Algebra R A] [Module R M₁] [Module A M₁]
+variable [SMulCommClass R A M₁] [SMulCommClass A R M₁] [IsScalarTower R A M₁]
+variable [Module R M₂]
+
+/-- The tensor product of two bilinear forms injects into bilinear forms on tensor products.
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/BilinearForm/TensorProduct.lean:40]:
```suggestion
variable (R A) in
/-- The tensor product of two bilinear forms injects into bilinear forms on tensor products.
```

File: Mathlib/LinearAlgebra/Dual.lean [hunk 2]
```diff
@@ -1608,6 +1609,25 @@ theorem dualDistrib_apply (f : Dual R M) (g : Dual R N) (m : M) (n : N) :
 
 end
 
+namespace AlgebraTensorModule
+set_option autoImplicit false
```
Comment (ocfnash) [inline Mathlib/LinearAlgebra/Dual.lean:1613]:
```suggestion

```

**Review by ocfnash — APPROVED** [review]
Thanks!

bors d+

**Comment by eric-wieser** [comment]
bors merge

**Pull request closed without merging**